bundle/edu.gemini.pot/src/test/scala/edu/gemini/spModel/target/env/GuideGroupSpec.scala

```diff
@@ -177,90 +175,92 @@ class GuideGroupSpec extends Specification with ScalaCheck with Arbitraries {
-      forAll { (g: GuideGroup) =>
-        true
-      }
-
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
-      }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup putAll" should {
+    "contain the original primary guide probes and all primary guide probes for which targets have been added" in
+      forAll { (g: GuideGroup, gps: Set[GuideProbe]) =>
+        val gpts = gps.map(GuideProbeTargets.create(_, new SPTarget())).toList.asImList
+        val allGps = g.getPrimaryReferencedGuiders.asScala.toSet ++ gps
+        g.putAll(gpts).getPrimaryReferencedGuiders.asScala.toSet == allGps
       }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
-      }
   }
 
-  "GuideGroup getAll" should {
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup setAll" should {
+    "only contain the guide probes in the collection" in
+      forAll { (g: GuideGroup, gps: Set[GuideProbe]) =>
+        val gpts = gps.map(GuideProbeTargets.create(_, new SPTarget())).toList.asImList
+        val g2 = g.setAll(gpts)
+        val gpOut = g.getReferencedGuiders.asScala.toSet -- g2.getReferencedGuiders.asScala.toSet
+        g2.getReferencedGuiders.asScala.toSet == gpts.asScalaList.map(_.getGuider).toSet &&
+          gpOut.forall(gp => !g2.contains(gp))
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup getPrimaryReferencedGuiders" should {
+    "contain a guider iff it has a primary target" in
+      forAll { g: GuideGroup =>
+        g.getAll.asScalaList.collect {
+          case gpt if gpt.getPrimary.isDefined => gpt.getGuider
+        }.toSet == g.getPrimaryReferencedGuiders.asScala.toSet
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup getReferencedGuiders" should {
+    "contain a guider iff it has associated targets" in
+      forAll { g: GuideGroup =>
+        g.getAll.asScalaList.map(_.getGuider).toSet == g.getReferencedGuiders.asScala.toSet
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup removeTarget" should {
+    "remove an existing target added to all guiders and preserve the order of the original targets" in
+      forAll { (g: GuideGroup, target: SPTarget) =>
+        val oldGpts = g.getAll
+        val newGpts = g.getAll.asScalaList.map(gt => gt.update(OptionsList.UpdateOps.append(target))).asImList
+        val remGpts = g.setAll(newGpts).removeTarget(target).getAll
+        remGpts == oldGpts
       }
+  }
 
-    "" in
-      forAll { (g: GuideGroup) =>
-        true
+  "GuideGroup getAllMatching" should {
+    "partition the collections of guide probe targets by guide probe type" in
+      forAll { g: GuideGroup =>
+        val ptns = g.getReferencedGuiders.asScala.map(gp => g.getAllMatching(gp.getType).asScalaList.toSet).toSet
+        (ptns.toList.map(_.length).sum == g.getAll.size) && ptns.forall(p => ptns.forall(q => p == q || p.intersect(q).isEmpty))
```

Comment:
I'm not quite following the logic of this one.  Let's add an `Arbitrary` for guide probe type and then add a `GuideProbeType` argument to `forAll` to get a random one.  Then check that all `GuideProbeTargets` returned by `getAllMatching` have a guide probe of that type.  Also, that `getAll` doesn't return anything with that type not in `getAllMatching`.

So, maybe something roughly like:

````scala
forAll { (g: GuideGroup, t: GuideProbeType) =>
  val all     = g.getAll.asScalaList.sortBy(_.getGuideProbe)
  val matches = g.getAllMatching(t).asScalaList.sortBy(_.getGuideProbe)

  all.filter(_.getGuideProbe.getType == t) == matches  
}
````